### ISSUE TITLE
PHPStorm plug-in workshop

### ISSUE THREAD

**Issue by garethellis36** [phpoxford/ideas#2]
Just noting down some ideas for a plugin I want to develop out of @ascii-soup 's proposed work-shop.

CakePHP plugin (update [CakeStorm](https://github.com/vexus2/CakeStorm)?)
- Add auto-complete support for `App::uses()` calls
- Similarly, add support for automatically adding an `App::uses()` call to top of PHP file when a class is instantiated or called statically (similar to how PHPStorm does this with vanilla PHP `use`)
- Add intention to add a docblock property for associated models (see below); ditto for auto-loaded model in Controller
- Add a view file from a controller. Perhaps via ALT+ENTER. PHPStorm would prompt user for view file name and then know which folder to save it in based on the controller name (e.g. `app/Controller/UsersController.php` would have views in `app/Views/Users/`. If it could create the file from a file template that would be even better. Alternatively to ALT+ENTER, current CakeStorm has shortcut `CTRL+ ; s` (ctrl and semi-colon at same time, then 's') which gives quick-jump options from Controller to its view files. A shortcut could be added here to create a new view file.
- Smart-jump for elements/view files, e.g.:

Pressing ctrl-B on the string "element_name" should open `app/View/Elements/element_name.ctp`.

``` html
//in a view file
<?= $this->Element("element_name") ?>
```

Pressing ctrl-B on the string "view_file" should open `app/View/Users/view.ctp`:

``` php
//UsersController.php
$this->render("view");
```

**Comment by ascii-soup** [comment]
We could probably add an Inspection for classes that have been used without
an App::uses() call and add an intention (alt+enter) to add the appropriate
cake import. Not sure if that would be enough

**Comment by garethellis36** [comment]
Example for 3rd bullet:

``` php
<?php

/**
 * Class ClientsContact
 * 
 * @property Client $Client
 * @property Region $Region
 */
class ClientsContact extends AppModel
{
    public $belongsTo = [
        "Region",
        "Client",
    ];
}
```

With cursor on the `$belongsTo` field, there should be an intention to add it as a docblock property. The same applies to the other model associations (`hasMany`, `hasOne` and `hasAndBelongsToMany`).

**Comment by ascii-soup** [comment]
Can I get an idea of numbers please - anyone who wants to attend could you reply for me?
Also, need to start thinking about dates. I'd like it to be on a Saturday so we can have the whole day (plus some pub time at the end) and Gareth has requested that this be before May.

If we start looking at April, how are people fixed for the 2nd, the 9th and the 16th?

Also, we'd need a venue. @olorton is this something White October would be willing to provide?

**Comment by garethellis36** [comment]
:+1:

All those dates are fine for me. The only Saturday in April I couldn't make is the 30th.

**Comment by garethellis36** [comment]
If WO doesn't work, I can ask if we can charge a meeting room at the Milton Park business centre to my company...

**Comment by ascii-soup** [comment]
If that doesn't work, how much do those meeting rooms go for?

**Comment by ascii-soup** [comment]
A fair bit, by the looks of it: £125 - £230

http://www.mepc.com/miltonpark/conferencesandmeetings/meetingsandconferences.aspx

**Comment by garethellis36** [comment]
Yeah £125 for a 6-person roundtable or £230 for 12/15 person boadroom. Bigger options also available.

**Comment by ascii-soup** [comment]
Well I'd like to keep the numbers to a minimum really..!

**Comment by liquorvicar** [comment]
2nd and 9th are best for me. If WO can't do it, I could ask Elsevier if we could do something there...

**Comment by ascii-soup** [comment]
To be fair, WGT would probably let us do it here, but the projector is pretty low res.

**Comment by garethellis36** [comment]
I've asked my boss if we can "sponsor" it. I've pitched it as a training cost, shouldn't be an issue in monetary terms.

**Comment by ascii-soup** [comment]
:+1:

**Comment by garethellis36** [comment]
No problem for us to sponsor it. Can probably get food & drinks paid for as well.

**Comment by garethellis36** [comment]
Shall I set-up a Doodle for collecting availability dates?

**Comment by ascii-soup** [comment]
Good idea yeah

**Comment by garethellis36** [comment]
http://doodle.com/poll/4a7cdcv8f67r4zga

**Comment by olorton** [comment]
I'm fairly sure that WO can host this, I'll need to be there obviously (unless anyone at WO is interested in this).

I cannot do the 2nd April, but the 9th or the 16th works for me.

**Comment by olorton** [comment]
@garethellis36 Assuming that it is not on the 2nd, would you still be happy to get some sponsorship sorted for drinks/food (e.g. a trip to Ron's Burrito Shack!)

**Comment by garethellis36** [comment]
Yes that would be fine.

**Comment by garethellis36** [comment]
Valid in PHP5:

``` php
$arr = array(1,2);
foreach ($arr as &$value) {
  $value++;
}
//[2,3]
```

In PHP7 the original $arr is unchanged. 

Maybe a plugin could be written to add an inspection to find instances of this (existing PHP7 inspection doesn't cover this) and even better fix as follows:

``` php
$arr = array(1,2);
foreach ($arr as $key => $value) {
  $arr[$key]++;
}
```

**Comment by ascii-soup** [comment]
That could be possible, but might be difficult to catch all cases where this happens. We can investigate!

Would be nice to have an option to replace with array_map() instead:

``` php
$arr = [1, 2];
$arr = array_map(function($value) {  return $value++; }, $arr);
```

**Comment by garethellis36** [comment]
Having investigated this on 3v4l, it turns out PHP7 doesn't handle this differently after all!

**Comment by garethellis36** [comment]
Looks like 9th April works for everyone. @olorton can you confirm that we can use WO please? If not I'll get another venue booked.

**Comment by garethellis36** [comment]
Bump @olorton
